File: backend/src/models/Payment.js
const mongoose = require('mongoose');

const paymentSchema = new mongoose.Schema({
  paymentId: {
    type: String,
    unique: true,
    required: true
  },
  booking: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Booking',
    required: [true, 'Booking reference is required']
  },
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'User reference is required']
  },
  property: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Property',
    required: [true, 'Property reference is required']
  },
  host: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Host reference is required']
  },
  amount: {
    total: {
      type: Number,
      required: [true, 'Total amount is required'],
      min: [0, 'Amount cannot be negative']
    },
    baseAmount: { type: Number, required: true },
    platformFee: { type: Number, required: true },
    hostAmount: { type: Number, required: true },
    taxAmount: { type: Number, default: 0 },
    currency: { type: String, default: 'USD', enum: ['USD', 'EUR', 'GBP', 'CAD', 'AUD'] }
  },
  method: {
    type: {
      type: String,
      enum: ['credit_card', 'debit_card', 'paypal', 'bank_transfer', 'stripe', 'apple_pay', 'google_pay'],
      required: [true, 'Payment method is required']
    },
    cardDetails: {
      last4: String,
      brand: String,
      expiryMonth: Number,
      expiryYear: Number
    },
    bankDetails: {
      accountLast4: String,
      bankName: String
    }
  },
  status: {
    type: String,
    enum: ['pending', 'processing', 'completed', 'failed', 'cancelled', 'refunded', 'partially_refunded'],
    default: 'pending'
  },
  transaction: {
    stripePaymentIntentId: String,
    stripeChargeId: String,
    paypalTransactionId: String,
    bankTransactionId: String,
    transactionFee: { type: Number, default: 0 },
    netAmount: { type: Number, required: true },
    exchangeRate: { type: Number, default: 1 }
  },
  timeline: {
    initiatedAt: { type: Date, default: Date.now },
    processedAt: Date,
    completedAt: Date,
    failedAt: Date,
    cancelledAt: Date,
    refundedAt: Date
  },
  failure: {
    reason: String,
    code: String,
    message: String,
    retryAttempts: { type: Number, default: 0 },
    maxRetries: { type: Number, default: 3 }
  },
  refund: {
    amount: { type: Number, default: 0 },
    reason: String,
    processedBy: {
      type: String,
      enum: ['guest', 'host', 'admin', 'system']
    },
    refundMethod: String,
    stripeRefundId: String,
    paypalRefundId: String,
    bankRefundId: String,
    estimatedRefundDate: Date,
    actualRefundDate: Date
  },
  metadata: {
    ipAddress: String,
    userAgent: String,
    deviceType: String,
    browserInfo: String,
    referrer: String
  },
  webhooks: [{
    eventType: String,
    eventId: String,
    receivedAt: { type: Date, default: Date.now },
    processed: { type: Boolean, default: false },
    data: mongoose.Schema.Types.Mixed
  }],
  notes: String
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Virtual for payment status display
paymentSchema.virtual('statusDisplay').get(function() {
  const statusMap = {
    'pending': 'Pending',
    'processing': 'Processing',
    'completed': 'Completed',
    'failed': 'Failed',
    'cancelled': 'Cancelled',
    'refunded': 'Refunded',
    'partially_refunded': 'Partially Refunded'
  };
  return statusMap[this.status] || this.status;
});

// Virtual for method display
paymentSchema.virtual('methodDisplay').get(function() {
  const methodMap = {
    'credit_card': 'Credit Card',
    'debit_card': 'Debit Card',
    'paypal': 'PayPal',
    'bank_transfer': 'Bank Transfer',
    'stripe': 'Stripe',
    'apple_pay': 'Apple Pay',
    'google_pay': 'Google Pay'
  };
  return methodMap[this.method.type] || this.method.type;
});

// Indexes for better query performance
paymentSchema.index({ paymentId: 1 });
paymentSchema.index({ user: 1, createdAt: -1 });
paymentSchema.index({ host: 1, createdAt: -1 });
paymentSchema.index({ booking: 1 });
paymentSchema.index({ status: 1 });
paymentSchema.index({ 'transaction.stripePaymentIntentId': 1 });
paymentSchema.index({ 'timeline.completedAt': -1 });

// Pre-save middleware to generate payment ID
paymentSchema.pre('save', function(next) {
  if (this.isNew && !this.paymentId) {
    const timestamp = Date.now().toString().slice(-6);
    const random = Math.floor(Math.random() * 1000).toString().padStart(3, '0');
    this.paymentId = `PAY${timestamp}${random}`;
  }
  
  // Calculate net amount
  if (this.isModified('amount.total') || this.isModified('transaction.transactionFee')) {
    this.transaction.netAmount = this.amount.total - (this.transaction.transactionFee || 0);
  }
  
  next();
});

// Static method to get payment stats
paymentSchema.statics.getPaymentStats = async function() {
  const stats = await this.aggregate([
    {
      $group: {
        _id: null,
        totalPayments: { $sum: 1 },
        totalAmount: { $sum: '$amount.total' },
        pendingPayments: { $sum: { $cond: [{ $eq: ['$status', 'pending'] }, 1, 0] } },
        completedPayments: { $sum: { $cond: [{ $eq: ['$status', 'completed'] }, 1, 0] } },
        failedPayments: { $sum: { $cond: [{ $eq: ['$status', 'failed'] }, 1, 0] } },
        refundedPayments: { $sum: { $cond: [{ $in: ['$status', ['refunded', 'partially_refunded']] }, 1, 0] } },
        averageTransactionValue: { $avg: '$amount.total' }
      }
    }
  ]);
  
  // Calculate monthly revenue
  const currentMonth = new Date();
  currentMonth.setDate(1);
  currentMonth.setHours(0, 0, 0, 0);
  
  const monthlyRevenue = await this.aggregate([
    {
      $match: {
        status: 'completed',
        'timeline.completedAt': { $gte: currentMonth }
      }
    },
    {
      $group: {
        _id: null,
        revenue: { $sum: '$amount.total' }
      }
    }
  ]);
  
  return {
    totalPayments: stats[0]?.totalPayments || 0,
    totalAmount: stats[0]?.totalAmount || 0,
    pendingPayments: stats[0]?.pendingPayments || 0,
    completedPayments: stats[0]?.completedPayments || 0,
    failedPayments: stats[0]?.failedPayments || 0,
    refundedPayments: stats[0]?.refundedPayments || 0,
    averageTransactionValue: Math.round(stats[0]?.averageTransactionValue || 0),
    monthlyRevenue: monthlyRevenue[0]?.revenue || 0
  };
};

// Method to process refund
paymentSchema.methods.processRefund = function(refundAmount, reason, processedBy) {
  if (this.status !== 'completed') {
    throw new Error('Only completed payments can be refunded');
  }
  
  if (refundAmount > this.amount.total) {
    throw new Error('Refund amount cannot exceed original payment amount');
  }
  
  const totalRefunded = this.refund.amount + refundAmount;
  
  if (totalRefunded > this.amount.total) {
    throw new Error('Total refund amount cannot exceed original payment amount');
  }
  
  this.refund.amount = totalRefunded;
  this.refund.reason = reason;
  this.refund.processedBy = processedBy;
  this.refund.actualRefundDate = new Date();
  
  if (totalRefunded === this.amount.total) {
    this.status = 'refunded';
    this.timeline.refundedAt = new Date();
  } else {
    this.status = 'partially_refunded';
  }
  
  return this.save();
};

// Method to update payment status
paymentSchema.methods.updateStatus = function(newStatus, additionalData = {}) {
  const statusTimelineMap = {
    'processing': 'processedAt',
    'completed': 'completedAt',
    'failed': 'failedAt',
    'cancelled': 'cancelledAt',
    'refunded': 'refundedAt'
  };
  
  this.status = newStatus;
  
  if (statusTimelineMap[newStatus]) {
    this.timeline[statusTimelineMap[newStatus]] = new Date();
  }
  
  // Update failure information if status is failed
  if (newStatus === 'failed' && additionalData.failure) {
    this.failure = {
      ...this.failure,
      ...additionalData.failure
    };
  }
  
  return this.save();
};

module.exports = mongoose.model('Payment', paymentSchema);

File: admin-dashboard/lib/models/payment.dart
import 'package:equatable/equatable.dart';

enum PaymentStatus {
  pending,
  completed,
  failed,
  refunded,
  cancelled,
}

enum PaymentMethod {
  creditCard,
  debitCard,
  paypal,
  bankTransfer,
  stripe,
  applePay,
  googlePay,
}

class Payment extends Equatable {
  const Payment({
    required this.id,
    required this.bookingId,
    required this.userId,
    required this.amount,
    required this.currency,
    required this.status,
    required this.paymentMethod,
    required this.transactionId,
    required this.paymentDate,
    this.feeAmount,
    this.refundAmount,
    this.refundDate,
    this.notes,
  });

  final String id;
  final String bookingId;
  final String userId;
  final double amount;
  final String currency;
  final PaymentStatus status;
  final PaymentMethod paymentMethod;
  final String transactionId;
  final DateTime paymentDate;
  final double? feeAmount;
  final double? refundAmount;
  final DateTime? refundDate;
  final String? notes;

  Payment copyWith({
    String? id,
    String? bookingId,
    String? userId,
    double? amount,
    String? currency,
    PaymentStatus? status,
    PaymentMethod? paymentMethod,
    String? transactionId,
    DateTime? paymentDate,
    double? feeAmount,
    double? refundAmount,
    DateTime? refundDate,
    String? notes,
  }) {
    return Payment(
      id: id ?? this.id,
      bookingId: bookingId ?? this.bookingId,
      userId: userId ?? this.userId,
      amount: amount ?? this.amount,
      currency: currency ?? this.currency,
      status: status ?? this.status,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      transactionId: transactionId ?? this.transactionId,
      paymentDate: paymentDate ?? this.paymentDate,
      feeAmount: feeAmount ?? this.feeAmount,
      refundAmount: refundAmount ?? this.refundAmount,
      refundDate: refundDate ?? this.refundDate,
      notes: notes ?? this.notes,
    );
  }

  String get statusText {
    switch (status) {
      case PaymentStatus.pending:
        return 'Pending';
      case PaymentStatus.completed:
        return 'Completed';
      case PaymentStatus.failed:
        return 'Failed';
      case PaymentStatus.refunded:
        return 'Refunded';
      case PaymentStatus.cancelled:
        return 'Cancelled';
    }
  }

  String get methodText {
    switch (paymentMethod) {
      case PaymentMethod.creditCard:
        return 'Credit Card';
      case PaymentMethod.debitCard:
        return 'Debit Card';
      case PaymentMethod.paypal:
        return 'PayPal';
      case PaymentMethod.bankTransfer:
        return 'Bank Transfer';
      case PaymentMethod.stripe:
        return 'Stripe';
      case PaymentMethod.applePay:
        return 'Apple Pay';
      case PaymentMethod.googlePay:
        return 'Google Pay';
    }
  }

  @override
  List<Object?> get props => [
        id,
        bookingId,
        userId,
        amount,
        currency,
        status,
        paymentMethod,
        transactionId,
        paymentDate,
        feeAmount,
        refundAmount,
        refundDate,
        notes,
      ];

  // Sample data
  static List<Payment> samplePayments = [
    Payment(
      id: 'PAY001',
      bookingId: 'B001',
      userId: 'U001',
      amount: 1200.00,
      currency: 'USD',
      status: PaymentStatus.completed,
      paymentMethod: PaymentMethod.creditCard,
      transactionId: 'TXN_123456789',
      paymentDate: DateTime.now().subtract(const Duration(days: 2)),
      feeAmount: 36.00,
    ),
    Payment(
      id: 'PAY002',
      bookingId: 'B002',
      userId: 'U002',
      amount: 850.00,
      currency: 'USD',
      status: PaymentStatus.pending,
      paymentMethod: PaymentMethod.paypal,
      transactionId: 'TXN_987654321',
      paymentDate: DateTime.now().subtract(const Duration(hours: 5)),
      feeAmount: 25.50,
    ),
    Payment(
      id: 'PAY003',
      bookingId: 'B003',
      userId: 'U003',
      amount: 1500.00,
      currency: 'USD',
      status: PaymentStatus.refunded,
      paymentMethod: PaymentMethod.stripe,
      transactionId: 'TXN_456789123',
      paymentDate: DateTime.now().subtract(const Duration(days: 7)),
      feeAmount: 45.00,
      refundAmount: 1500.00,
      refundDate: DateTime.now().subtract(const Duration(days: 1)),
      notes: 'Guest requested cancellation due to travel restrictions.',
    ),
    Payment(
      id: 'PAY004',
      bookingId: 'B004',
      userId: 'U004',
      amount: 900.00,
      currency: 'USD',
      status: PaymentStatus.failed,
      paymentMethod: PaymentMethod.bankTransfer,
      transactionId: 'TXN_789123456',
      paymentDate: DateTime.now().subtract(const Duration(hours: 12)),
      feeAmount: 27.00,
      notes: 'Insufficient funds in bank account.',
    ),
    Payment(
      id: 'PAY005',
      bookingId: 'B005',
      userId: 'U005',
      amount: 2100.00,
      currency: 'USD',
      status: PaymentStatus.completed,
      paymentMethod: PaymentMethod.applePay,
      transactionId: 'TXN_321654987',
      paymentDate: DateTime.now().subtract(const Duration(days: 1)),
      feeAmount: 63.00,
    ),
  ];
}

class PaymentStats {
  const PaymentStats({
    required this.totalPayments,
    required this.totalAmount,
    required this.pendingPayments,
    required this.completedPayments,
    required this.failedPayments,
    required this.refundedPayments,
    required this.averageTransactionValue,
    required this.monthlyRevenue,
  });

  final int totalPayments;
  final double totalAmount;
  final int pendingPayments;
  final int completedPayments;
  final int failedPayments;
  final int refundedPayments;
  final double averageTransactionValue;
  final double monthlyRevenue;

  static PaymentStats calculateStats(List<Payment> payments) {
    final totalPayments = payments.length;
    final totalAmount = payments.fold(0.0, (sum, payment) => sum + payment.amount);
    final pendingPayments = payments.where((p) => p.status == PaymentStatus.pending).length;
    final completedPayments = payments.where((p) => p.status == PaymentStatus.completed).length;
    final failedPayments = payments.where((p) => p.status == PaymentStatus.failed).length;
    final refundedPayments = payments.where((p) => p.status == PaymentStatus.refunded).length;
    final averageTransactionValue = totalPayments > 0 ? totalAmount / totalPayments : 0.0;
    
    final currentMonth = DateTime.now().month;
    final currentYear = DateTime.now().year;
    final monthlyRevenue = payments
        .where((p) => p.paymentDate.month == currentMonth && 
                     p.paymentDate.year == currentYear &&
                     p.status == PaymentStatus.completed)
        .fold(0.0, (sum, payment) => sum + payment.amount);

    return PaymentStats(
      totalPayments: totalPayments,
      totalAmount: totalAmount,
      pendingPayments: pendingPayments,
      completedPayments: completedPayments,
      failedPayments: failedPayments,
      refundedPayments: refundedPayments,
      averageTransactionValue: averageTransactionValue,
      monthlyRevenue: monthlyRevenue,
    );
  }
}
